src/filesystem/index.ts

```diff
@@ -202,6 +216,61 @@ async function searchFiles(
   return results;
 }
 
+// Edit preview type
+interface EditPreview {
+  original: string;
+  modified: string;
+  lineNumber: number;
+  preview: string;  // Git-style diff format
+}
+
+// File editing utilities
+async function applyFileEdits(filePath: string, edits: z.infer<typeof EditOperation>[]): Promise<string | EditPreview[]> {
+  let content = await fs.readFile(filePath, 'utf-8');
+  const previews: EditPreview[] = [];
+  
+  for (const edit of edits) {
+    const pos = content.indexOf(edit.oldText);
+    if (pos === -1) {
+      throw new Error(
+        `Search text not found in ${filePath}:\n${edit.oldText}`
+      );
+    }
+    
+    // Calculate line number for reporting
+    const lineNumber = content.slice(0, pos).split(/\r?\n/).length;
+    
+    if (edit.dryRun) {
+      // Create git-style diff preview
+      const preview = [
+        `@@ line ${lineNumber} @@`,
+        '<<<<<<< ORIGINAL',
+        edit.oldText,
+        '=======',
+        edit.newText,
+        '>>>>>>> MODIFIED'
+      ].join('\n');
+      
+      previews.push({
+        original: edit.oldText,
+        modified: edit.newText,
+        lineNumber,
+        preview
+      });
+      continue;
+    }
+    
+    // Apply the edit
+    content = content.slice(0, pos) + edit.newText + content.slice(pos + edit.oldText.length);
+  }
+  
+  if (edits.some(e => e.dryRun)) {
+    return previews;
+  }
```

Comment:
Since `dryRun` seems changes the logic _and_ the return type, can we create two different functions instead? We can reuse common bits between them instead of `if`/`else`-ing most of the logic in one function.

Reply:
Looks like this is still a good suggestion after the refactoring I did last night. I can take a look at this later as I'll need to rerun all of my tests after making this change.